lib/kafo/validator.rb

```diff
@@ -2,79 +2,111 @@
+        unless arg.to_s.start_with?('/')
+          @logger.error "#{arg.inspect} is not an absolute path."
+          return false
+        end
       end
+      return true
+    end
 
-      @params = params
-      @logger = KafoConfigure.logger
+    def validate_array(args)
+      args.each do |arg|
+        unless arg.is_a?(Array)
+          @logger.error "#{arg.inspect} is not a valid array."
+          return false
+        end
+      end
+      return true
+    end
 
-      @cache ||= Hash.new do |hash, key|
-        @logger.debug "Looked for #{key}"
-        param     = @params.select { |p| p.name == key.to_s }.first
-        hash[key] = param.nil? ? nil : param.value
+    def validate_bool(args)
+      args.each do |arg|
+        unless arg.is_a?(TrueClass) || arg.is_a?(FalseClass)
+          @logger.error "#{arg.inspect} is not a valid boolean."
+          return false
+        end
       end
+      return true
     end
 
-    def lookupvar(name, options = {})
-      @cache[name]
+    def validate_hash(args)
+      args.each do |arg|
+        unless arg.is_a?(Hash)
+          @logger.error "#{arg.inspect} is not a valid hash."
+          return false
+        end
+      end
+      return true
     end
 
-    # for puppet >= 3
-    def include?(value)
-      true
+    def validate_integer(args)
+      value = args[0]
+      max = args[1]
+      min = args[2]
+      int = Integer(value.to_s)
+      if min && int < min.to_i
+        @logger.error "#{value} must be at least #{min}."
+        return false
+      end
+      if max && int > max.to_i
+        @logger.error "#{value} must be less than #{max}."
+        return false
+      end
+      return true
+    rescue TypeError, ArgumentError
+      @logger.error "#{value.inspect} is not a valid integer."
+      return false
     end
 
-    # for puppet >= 3
-    def [](value, *args)
-      lookupvar(value)
+    def validate_listen_on(args)
+      valid_values = ['http', 'https', 'both']
+      args.each do |arg|
+        unless valid_values.include?(arg)
+          @logger.error "#{arg.inspect} is not a valid value.  Valid values are: #{valid_values.join(", ")}"
+          return false
+        end
+      end
+      return true
     end
 
-    def method_missing(method, *args, &block)
-      method.to_s =~ /^function_(.*)$/
-      super unless $1
-      super unless Puppet::Parser::Functions.function($1)
-      # In odd circumstances, this might not end up defined by the previous
-      # method, so we might as well be certain.
-      if engine.respond_to? method
-        @logger.debug "calling #{method.inspect} with #{args.inspect}"
-        engine.send(method, *args)
-        true
+    def validate_re(args)
+      value = args[0]
+      regexes = args[1]
+      regexes = [regexes] unless regexes.is_a?(Array)
+      message = args[2] || "#{value.inspect} does not match the accepted inputs: #{regexes.join(", ")}"
+
+      if regexes.all? { |rx| value =~ Regexp.compile(rx) }
```

Comment:
when args[1] is a `Puppet::Parser::AST::ASTArray` it gets wrapped by `[]` and on this line fails with `initialize': 

```
[root@installertest kafo-0.7.3]# foreman-installer --foreman-email-delivery-method wrong             
/usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/validator.rb:85:in `initialize': no implicit conversion of Puppet::Parser::AST::ASTArray into String (TypeError)
        from /usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/validator.rb:85:in `compile'
        from /usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/validator.rb:85:in `block in validate_re'
        from /usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/validator.rb:85:in `each'
        from /usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/validator.rb:85:in `all?'
        from /usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/validator.rb:85:in `validate_re'
        from /usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/param.rb:90:in `block in valid?'
        from /usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/param.rb:90:in `each'
        from /usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/param.rb:90:in `all?'
        from /usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/param.rb:90:in `valid?'
        from /usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/kafo_configure.rb:365:in `block in validate_all'
        from /usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/kafo_configure.rb:364:in `map'
        from /usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/kafo_configure.rb:364:in `validate_all'
        from /usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/kafo_configure.rb:135:in `execute'
        from /usr/share/gems/gems/clamp-1.0.0/lib/clamp/command.rb:68:in `run'
        from /usr/share/gems/gems/clamp-1.0.0/lib/clamp/command.rb:133:in `run'
        from /usr/local/share/gems/gems/kafo-0.7.3/lib/kafo/kafo_configure.rb:156:in `run'
        from /usr/sbin/foreman-installer:12:in `<main>'
```

to fix it I had to add check on line 82 for this class and call `Regexp.compile(rx.to_s)` on this line, since ASTArray contains ASTString which would fail with similar error